include/exadg/compressible_navier_stokes/user_interface/application_base.h

```diff
@@ -60,39 +61,76 @@ class ApplicationBase
     // clang-format on
   }
 
-  ApplicationBase(std::string parameter_file)
-    : parameter_file(parameter_file), n_subdivisions_1d_hypercube(1)
+  ApplicationBase(std::string parameter_file, MPI_Comm const & comm)
+    : mpi_comm(comm), parameter_file(parameter_file), n_subdivisions_1d_hypercube(1)
   {
   }
 
   virtual ~ApplicationBase()
   {
   }
 
+  void
+  setup(unsigned int const         degree,
+        unsigned int const         refine_space,
+        ConditionalOStream const & pcout)
+  {
+    parameters.degree = degree;
+
+    set_input_parameters();
+    parameters.check_input_parameters();
+    parameters.print(pcout, "List of input parameters:");
+
+    // grid
+    grid_data.triangulation_type = parameters.triangulation_type;
+    grid_data.n_refine_global    = refine_space;
+    grid_data.mapping_degree     = get_mapping_degree(parameters.mapping, parameters.degree);
+
+    create_grid();
+    print_grid_info(pcout, *grid);
+
+    boundary_descriptor = std::make_shared<BoundaryDescriptor<dim>>();
+    field_functions     = std::make_shared<FieldFunctions<dim>>();
+
+    set_boundary_conditions();
+    CompNS::verify_boundary_conditions<dim>(boundary_descriptor, *grid);
+
+    set_field_functions();
+  }
+
+private:
   virtual void
-  set_input_parameters(InputParameters & parameters) = 0;
+  set_input_parameters() = 0;
 
-  virtual std::shared_ptr<Grid<dim, Number>>
-  create_grid(GridData const & data, MPI_Comm const & mpi_comm) = 0;
+  virtual void
+  create_grid() = 0;
 
   virtual void
-  set_boundary_conditions(std::shared_ptr<BoundaryDescriptor<dim>> boundary_descriptor) = 0;
+  set_boundary_conditions() = 0;
 
   virtual void
-  set_field_functions(std::shared_ptr<FieldFunctions<dim>> field_functions) = 0;
+  set_field_functions() = 0;
 
+public:
   virtual std::shared_ptr<PostProcessorBase<dim, Number>>
-  create_postprocessor(unsigned int const degree, MPI_Comm const & mpi_comm) = 0;
+  create_postprocessor() = 0;
 
   void
   set_subdivisions_hypercube(unsigned int const n_subdivisions_1d)
   {
     n_subdivisions_1d_hypercube = n_subdivisions_1d;
   }
 
+  MPI_Comm const & mpi_comm;
+
+  InputParameters                          parameters;
+  std::shared_ptr<BoundaryDescriptor<dim>> boundary_descriptor;
+  std::shared_ptr<FieldFunctions<dim>>     field_functions;
+  GridData                                 grid_data;
+  std::shared_ptr<Grid<dim, Number>>       grid;
```

Comment:
I would make these protected so that the derived classes can access them directly but not the driver. The driver could get these via getter functions.